### DISCUSSION TITLE
Development strategy

### DISCUSSION THREAD

**Discussion by FaBjE** [spencerwmiles/vscode-task-buttons#35] [General]
_I think this is a "proper place" to discuss this, instead of a closed pull request 😉_

> @FaBjE
> 
> As you know, it has been a while since I've had the time to look things through here correctly. I had a full-rewrite of the extension sitting around, and yes, it's finally made its way out as v1.1.0. Unfortunately, @gameguy682 Quick Pick Menu was also left sitting until this point. Everything _should_ be caught up now.
> 
> I hope you agree it's a little cleaner and easier to work with and should be a much better foundation for us all to build from.
> 
> Your contributions to this project have meant a lot to me, and I have given you contributor access for that reason. Not only that, but three contributors make it way easier to "democratize" things and have some things hopefully getting to this point again.
> 
> At this point, the existing "branching strategy" is overly complicated for its value. Also, sorting things out after being away for as long as I did wasn't fun (even with all your help).
> 
> Right now, I plan to work towards a single production branch with auto-release via GitHub Actions.
> 
> I hope that helps a bit, my friend. Please add your thoughts wherever you see the value!

Thanks for the info 👍🏻 
I agree it is a much better starting point than the "old" version.

About the branching model, I'm used to working with git-flow. It seems fitting for this project as well.
https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
I could see this work with me and @gameguy682 doing feature branches and the occasional develop branch.
And you as "Lead contributer" the merges to main/master/production whatever you like to call it. (Controlling the releases)

About the code, I will familiarize myself with the new setup. I hope you don't mind if I add some comment lines.
I know if you write the code yourself it all seems very logical, but adopting someone else's code always is a challenge.

**Comment by spencerwmiles** [comment]
@FaBjE I'll give that a read-over on the weekend, and we can decide shortly after that. 

Feel free to add any comments and documentation you would like. But I suggest you use the production branch until we sort this out. It makes my life easier, and I can use it now.

PS. I'm sorry for any merge fumbles that have also occurred. I am still getting the hang of using GitHub in this capacity.

Things will get better gotta continue being patient with me.

**Comment by FaBjE** [comment]
>  But I suggest you use the production branch until we sort this out.

I assume you mean as the code source, me committing directly on the production branch does not seem wise...

> I'm sorry for any merge fumbles that have also occurred. I am still getting the hang of using GitHub in this capacity.

No worries, me too 😆 

> Things will get better gotta continue being patient with me.

👍🏻 


P.S.
Git-flow is a "huge" branching model with all kinds of options.
Normally I use the "basic" version of:
1. A main/master/production branch (the latest release is on their. Merging to it is only done by the maintainer)
2. develop branch. All feature branches are merged to this one. Merging only after a pull-request is approved.
3. Feature branches (all start with "feature/") a "sandbox" for every feature.

**Comment by gameguy682** [comment]
Nothing wrong with Gitflow, and I believe that's how @spencerwmiles was thinking about doing it at first, but then the rewrite into typescript happened. As things are pretty small, it doesn't really matter if we keep things on the `production` branch or if we make a `devops` branch and do work from there.

Personally, I would like to do trunk based development (where you merge directly into `production` but have a flag to stop it being used), but it's hard without feature flags and I don't think we can add those to extensions (could be wrong, just didn't really look).

**Comment by spencerwmiles** [comment]
Unrelated, but we crossed 4k installs! 🎉

**Comment by spencerwmiles** [comment]
This is what i've got in place for the foreseeable future.

valhalla
          ⬆️
production      | releases
          ⬆️
staging         | prerelease / release candidates
          ⬆️
dev             | bugs, features

---

Every merge in any of these branches requires a PR with passing build. (until we reach valhalla)

A merge to staging will create a new  pre-release/release candidate in both GitHub & VSCode Marketplace.

A merge to production from staging will turn that pre-release into a full release.

Things should be much easier moving forward and we can just focus on the fun stuff.
